link/uprobe_multi.go

```diff
@@ -0,0 +1,230 @@
+
+	// We have either opts.Addresses array with absolute file offsets
+	// or we need to get from symbols array and optional opts.Offsets.
+	syms := len(symbols)
+	addrs := len(opts.Addresses)
+	offsets := len(opts.Offsets)
+
+	if addrs != 0 && (offsets != 0 || syms != 0) {
+		return fmt.Errorf("Addresses and Offsets/symbols are mutually exclusive: %w", errInvalidInput)
+	}
+
+	var err error
+	if syms != 0 {
+		// Translate symbols to addresses
+		opts.Addresses, err = ex.uprobeMultiAddresses(symbols, opts)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// addressMulti calculates the array of addresses from symbols array in the executable.
+//
+// opts must not be nil.
+func (ex *Executable) uprobeMultiAddresses(symbols []string, opts *UprobeMultiOptions) ([]uint64, error) {
+	var addresses []uint64
+
+	syms := len(symbols)
+	offsets := len(opts.Offsets)
+
+	// if Offsets array is defined, it must have same dimension as symbols
+	if offsets != 0 && offsets != syms {
+		return nil, fmt.Errorf("Offsets must be either zero or exactly symbols in length: %w", errInvalidInput)
+	}
+
+	for idx, symbol := range symbols {
+		address, err := ex.address(symbol)
+		if err != nil {
+			return nil, err
+		}
+		if offsets != 0 {
+			address += opts.Offsets[idx]
+		}
+		addresses = append(addresses, address)
+	}
+	return addresses, nil
+}
+
+type uprobeMultiLink struct {
+	RawLink
+}
+
+var _ Link = (*uprobeMultiLink)(nil)
+
+func (kml *uprobeMultiLink) Update(prog *ebpf.Program) error {
+	return fmt.Errorf("update uprobe_multi: %w", ErrNotSupported)
+}
+
+func (kml *uprobeMultiLink) Pin(string) error {
+	return fmt.Errorf("pin uprobe_multi: %w", ErrNotSupported)
+}
+
+func (kml *uprobeMultiLink) Unpin() error {
+	return fmt.Errorf("unpin uprobe_multi: %w", ErrNotSupported)
+}
+
+var haveBPFLinkUprobeMulti = internal.NewFeatureTest("bpf_link_uprobe_multi", "6.6", func() error {
+	prog, err := ebpf.NewProgram(&ebpf.ProgramSpec{
+		Name: "probe_upm_link",
+		Type: ebpf.Kprobe,
+		Instructions: asm.Instructions{
+			asm.Mov.Imm(asm.R0, 0),
+			asm.Return(),
+		},
+		AttachType: ebpf.AttachTraceUprobeMulti,
+		License:    "MIT",
+	})
+	if errors.Is(err, unix.E2BIG) {
+		// Kernel doesn't support AttachType field.
+		return internal.ErrNotSupported
+	}
+	if err != nil {
+		return err
+	}
+	defer prog.Close()
+
+	// We try to create uprobe multi link on '/' path which results in
+	// error with -EBADF in case uprobe multi link is supported.
+	fd, err := sys.LinkCreateUprobeMulti(&sys.LinkCreateUprobeMultiAttr{
+		ProgFd:     uint32(prog.FD()),
+		AttachType: sys.BPF_TRACE_UPROBE_MULTI,
+		Path:       sys.NewStringPointer("/"),
+		Offsets:    sys.NewPointer(unsafe.Pointer(&[]uint64{0})),
+		Count:      1,
+	})
+	switch {
+	case errors.Is(err, unix.EBADF):
+		return nil
+	case err != nil:
+		return internal.ErrNotSupported
+	}
+	// should not happen
+	fd.Close()
+	return internal.ErrNotSupported
```

Comment:
This somewhat violates the convention of our feature probe return values. `ErrNotSupported` means a conclusive 'unsupported' verdict, but successfully attaching a uprobe to `/` seems highly unexpected. I would return a bespoke `errors.New("successfully attached uprobe_multi to /, kernel bug?")` here to help the user. Also, we tend to return `ErrNotSupported` only on return codes we recognize:

```go
	switch {
	case errors.Is(err, unix.EBADF):
		return nil
	case errors.Is(err, unix.EINVAL):
		return internal.ErrNotSupported
	case err != nil:
		return err
	}

	fd.Close()
	return errors.New("successfully attached uprobe_multi to /, kernel bug?")
```

`kprobe_multi.go` also handles this case:
```
	// If CONFIG_FPROBE isn't set.
	case errors.Is(err, unix.EOPNOTSUPP):
		return internal.ErrNotSupported
```

Does that apply here as well?

Reply:
right, makes sense.. will change it, thanks